src/adc.rs

```diff
@@ -482,7 +482,9 @@ impl Buffer {
 
         // Let's translate what we got from the DMA peripheral into a write
         // position that we can compare with our read position.
-        let pos = self.len.saturating_sub(remaining) % self.len;
+        // Zero can be witnessed in remaining before the DMA reloads
+        // the register so we wrap the result of the subtraction around
+        let pos = (self.len - remaining) % self.len;
```

Comment:
Okay, so the thing I missed before is that if `remaining` is zero, the `% self.len` will result in `pos = 0` instead of `pos = self.len`. Makes sense.

What I don't understand yet is the wider context of this change. Under which circumstances will `remaining == 0` actually cause an overrun, and why? How will this affect overrun detection?

I'm not saying this is wrong, but I'm also not convinced of the following things:
- That this is correct, given the complexity of the surrounding code.
- That the overruns you're seeing are actually caused by `remaining == 0`, given the known issue with ADC/DMA.
- If you're experiencing overruns not caused by #104, that this is the right place to fix them. Maybe it should be fixed in the overrun detection code, for example.

This code is quite complex, and we're really suffering from the lack of a test suite here. I think the best way to get this merged is to do a more thorough analysis and write up the results in a comment in the appropriate place. As it is, I don't understand this fix (and its underlying problem) well enough to merge it.

Reply:
Yeah. I'll do that. I thought I had a good handle on it but I will review #104 and see if I can really pin this down

I think I misunderstood the bug but now looking at #104 I think this is a duplicate of that bug. Lets close this merge and my version of the bug and i'll focus my efforts there

Reply:
Sounds good!

Please note that my latest conclusion was, that this API can't be supported given the hardware behavior (https://github.com/stm32-rs/stm32l0xx-hal/issues/104#issuecomment-631383181), and I currently favor just adopting @apgoetz's new API to replace this one (https://github.com/stm32-rs/stm32l0xx-hal/pull/105#pullrequestreview-465789802).

But if you can come up with a fix for the current API, that would of course be great.